lib/ecmascript.ts

```diff
@@ -3232,444 +3360,468 @@ export const ES = ObjectAssign({}, ES2020, {
+    day: GetSlot(addedDate, ISO_DAY),
+    hour,
+    minute,
+    second,
+    millisecond,
+    microsecond,
+    nanosecond
+  };
+}
+
+export function AddZonedDateTime(
+  instant,
+  timeZone,
+  calendar,
+  years,
+  months,
+  weeks,
+  days,
+  h,
+  min,
+  s,
+  ms,
+  µs,
+  ns,
+  options?: any
+) {
+  // If only time is to be added, then use Instant math. It's not OK to fall
+  // through to the date/time code below because compatible disambiguation in
+  // the PlainDateTime=>Instant conversion will change the offset of any
+  // ZonedDateTime in the repeated clock time after a backwards transition.
+  // When adding/subtracting time units and not dates, this disambiguation is
+  // not expected and so is avoided below via a fast path for time-only
+  // arithmetic.
+  // BTW, this behavior is similar in spirit to offset: 'prefer' in `with`.
+  const TemporalDuration = GetIntrinsic('%Temporal.Duration%');
+  if (DurationSign(years, months, weeks, days, 0, 0, 0, 0, 0, 0) === 0) {
+    return AddInstant(GetSlot(instant, EPOCHNANOSECONDS), h, min, s, ms, µs, ns);
+  }
+
+  // RFC 5545 requires the date portion to be added in calendar days and the
+  // time portion to be added in exact time.
+  const dt = BuiltinTimeZoneGetPlainDateTimeFor(timeZone, instant, calendar);
+  const datePart = CreateTemporalDate(GetSlot(dt, ISO_YEAR), GetSlot(dt, ISO_MONTH), GetSlot(dt, ISO_DAY), calendar);
+  const dateDuration = new TemporalDuration(years, months, weeks, days, 0, 0, 0, 0, 0, 0);
+  const addedDate = CalendarDateAdd(calendar, datePart, dateDuration, options);
+  const dtIntermediate = CreateTemporalDateTime(
+    GetSlot(addedDate, ISO_YEAR),
+    GetSlot(addedDate, ISO_MONTH),
+    GetSlot(addedDate, ISO_DAY),
+    GetSlot(dt, ISO_HOUR),
+    GetSlot(dt, ISO_MINUTE),
+    GetSlot(dt, ISO_SECOND),
+    GetSlot(dt, ISO_MILLISECOND),
+    GetSlot(dt, ISO_MICROSECOND),
+    GetSlot(dt, ISO_NANOSECOND),
+    calendar
+  );
+
+  // Note that 'compatible' is used below because this disambiguation behavior
+  // is required by RFC 5545.
+  const instantIntermediate = BuiltinTimeZoneGetInstantFor(timeZone, dtIntermediate, 'compatible');
+  return AddInstant(GetSlot(instantIntermediate, EPOCHNANOSECONDS), h, min, s, ms, µs, ns);
+}
+
+export function RoundNumberToIncrement(quantity, increment, mode) {
+  if (increment === 1) return quantity;
+  let { quotient, remainder } = quantity.divmod(increment);
+  if (remainder.equals(bigInt.zero)) return quantity;
+  const sign = remainder.lt(bigInt.zero) ? -1 : 1;
+  switch (mode) {
+    case 'ceil':
+      if (sign > 0) quotient = quotient.add(sign);
+      break;
+    case 'floor':
+      if (sign < 0) quotient = quotient.add(sign);
+      break;
+    case 'trunc':
+      // no change needed, because divmod is a truncation
+      break;
+    case 'halfExpand':
+      // "half up away from zero"
+      if (remainder.multiply(2).abs() >= increment) quotient = quotient.add(sign);
```

Comment:
In addition to the zoneddatetime.ts bug described in https://github.com/js-temporal/temporal-polyfill/pull/57#discussion_r718042170, this line also looks like a bug if `increment` is a Number.

Reply:
https://github.com/peterolson/BigInteger.js/blob/4c1f2067768bfdd923b3d5e64392e90f1b748502/BigInteger.js#L1342-L1344

So long as that value isn't larger than Number.MAX_SAFE_INTEGER, this should be fine I think. This code will have to change once we use JSBI, as it doesn't do this (potentially lossy) conversion as all (yay!).

I'm going to return to this once we've migrated to JSBI, as we will likely have to deal with a bunch of these sorts of issues all at once.